graph-builder/src/graph.rs

```diff
@@ -296,15 +296,38 @@ pub fn run<'a>(settings: &'a config::AppSettings, state: &State) -> ! {
     }
 }
 
+/// Turns a collection of Releases into a Cincinnati Graph
+///
+/// When processing previous/next release metadata it is assumed that the edge
+/// destination has the same build type as the origin.
 pub fn create_graph(releases: Vec<registry::Release>) -> Result<Graph, Error> {
     let mut graph = Graph::default();
 
     releases
         .into_iter()
         .inspect(|release| trace!("Adding a release to the graph '{:?}'", release))
         .try_for_each(|release| {
-            let previous = release.metadata.previous.clone();
-            let next = release.metadata.next.clone();
+            let previous = release
+                .metadata
+                .previous
+                .iter()
+                .cloned()
+                .map(|mut previous| {
+                    previous.build = release.metadata.version.build.clone();
```

Comment:
I don't think we need this `build` mutation.  Can we ignore architecture when creating edges, link all versions regardless of `build`, and let the arch plugin filter any cross-arch edges out for us?

Reply:
> I don't think we need this `build` mutation.

This mutates the `build` of the previous (and further down next) edge declaration. The nodes in the graph they refer to are already arch-annotated through their build version. Without the mutation here we wouldn't find the correct edge endpoint. If an edge endpoint isn't found, [it creates an abstract release as a place-holder][abstract-create]. This is required because the insertion process has to handle the insertion of a node which declares edges to versions of nodes which are not yet inserted in the graph. If we skip the mutation discussed here, we'll end up with edges to abstract releases in the final graph, and fail to connect the edge to the desired annotated concrete releases.

[abstract-create]: https://github.com/steveeJ-forks/openshift-cincinnati/blob/0d5823ab6c58e25cd966d737ec1915ef5bbb0525/graph-builder/src/graph.rs#L336-L338

> Can we ignore architecture when creating edges, link all versions regardless of `build`, and let the arch plugin filter any cross-arch edges out for us?

IIUC that would be a different implementation of the same logic, where instead of creating arch-annotated release nodes, we'd have arch-annotated edges. The reason why I chose the former is simply because it was easier to do with the existing code.

Reply:
> Without the mutation here we wouldn't find the correct edge endpoint.

It wouldn't be found if you use `graph.find_by_version`.  But you'd find (potentially several) endpoints if you used a new `graph.iter_by_version(version: &str)` like I floated [here][1], right?

> This is required because the insertion process has to handle the insertion of a node which declares edges to versions of nodes which are not yet inserted in the graph.

Wait, why are we processing edges before we have added all the nodes?  Can't we slurp in the nodes (with whatever metadata, including edge requests), and then leave edge processing to the edge-add-remove plugin?

> IIUC that would be a different implementation of the same logic, where instead of creating arch-annotated release nodes, we'd have arch-annotated edges.

They wouldn't be annotated edges.  They'd be removed by the arch plugin because when it removed a non-matching node it would also remove all edges touching that node.

[1]: https://github.com/openshift/cincinnati/pull/176#discussion_r347750531

Reply:
>  But you'd find (potentially several) endpoints if you used a new `graph.iter_by_version(version: &str)` like I floated [here](https://github.com/openshift/cincinnati/pull/176#discussion_r347750531), right?

That entails finding and connecting releases of different architectures as the one which is currently looked at. That doesn't translate the input data correctly (i.e. it breaks) in cases where architectures don't have the same edge constellation.

Reply:
> That entails finding and connecting releases of different architectures as the one which is currently looked at.

Yup.  But that's fine, we'll filter mismatches out in the arch plugin when we drop the non-matching nodes.

> That doesn't translate the input data correctly (i.e. it breaks) in cases where architectures don't have the same edge constellation.

What's an edge constellation?

> I slightly changed the code to first add the releases and subsequently add the edges.

Is this 92f6cb3?  I think we can probably drop edge handling entirely except for the edge-add-remove plugin and removing associated edges whenever we remove a node.  But that doesn't have to happen in this PR.

Reply:
> Yup. But that's fine, we'll filter mismatches out in the arch plugin when we drop the non-matching nodes.

I'm not sure if it's fine. It's inaccurate, as it loses detail which the input data provides.



> What's an edge constellation?

What I meant by that is that, according to the input data, one architecture could have different edges than another architecture.



> Is this [92f6cb3](https://github.com/openshift/cincinnati/commit/92f6cb3099629b7d2656f25fe1942fd03be07e49)? I think we can probably drop edge handling entirely except for the edge-add-remove plugin and removing associated edges whenever we remove a node. But that doesn't have to happen in this PR.

That would require to copy all previous/next information from the release-metadata files into the release metadata which ends up in the graph, because that plugin doesn't have access to the former.

Reply:
> > But that's fine, we'll filter mismatches out in the arch plugin when we drop the non-matching nodes.
>
> I'm not sure if it's fine. It's inaccurate, as it loses detail which the input data provides.

But the input data does not contain arch-specific names, so we're not losing detail by creating edges to all possible versions regardless of arch.

> What I meant by that is that, according to the input data, one architecture could have different edges than another architecture.

I think that's fine, since it's temporary internal state.

> That would require to copy all previous/next information from the release-metadata files into the release metadata which ends up in the graph, because that plugin doesn't have access to the former.

That's fine, right?  I thought the idea was "we don't care if it was primary or secondary, it's all just metadata".

Reply:
> But the input data does not contain arch-specific names, so we're not losing detail by creating edges to all possible versions regardless of arch.

The images we scrape are arch specific thus all metadata, including its previous/next declarations, is specific to that arch. As we scrape all images of a given repository we add their arch-specific releases to the graph.

> I think that's fine, since it's temporary internal state.

I don't think it's fine, as the graph could end up with edges which weren't defined by any metadata.

> That's fine, right? I thought the idea was "we don't care if it was primary or secondary, it's all just metadata".

That should be fine. I suggest we handle that in a separate PR.